### PR TITLE
Rename SetNmAndUqTags and fix #622 (MergeBamAlignment MD tag).

### PR DESCRIPTION
**Pull request by nh13**
@yfarjoun @tfenne 

1. Rename SetNmAndUqTags to SetNmMDAndUqTags.
WIP
2. Fixes: https://github.com/broadinstitute/picard/pull/622

Fixes https://github.com/broadinstitute/picard/issues/634 as an alternate to reverting (https://github.com/broadinstitute/picard/pull/635).

### CODE REVIEW

File: src/main/java/picard/sam/AbstractAlignmentMerger.java [hunk 1]
```diff
@@ -585,20 +587,6 @@ private void transferAlignmentInfoToPairedRead(final SAMRecord firstUnaligned, f
...
-     * Updates the MD and NM tags if present.
-     * @param rec
-     */
-    private void calculateMdAndNmTags(final SAMRecord rec) {
-        // update NM and MD
-        final boolean calculateNm = rec.getIntegerAttribute(ReservedTagConstants.NM) != null;
-        final boolean calculateMd = rec.getStringAttribute(SAMTag.MD.name()) != null;
-        if (calculateNm || calculateMd) {
-            final byte[] referenceBases = this.refSeq.get(this.refSeq.getSequenceDictionary().getSequenceIndex(rec.getReferenceName())).getBases();
-            SequenceUtil.calculateMdAndNmTags(rec, referenceBases, calculateNm, calculateMd);
```
Comment (yfarjoun) [inline src/main/java/picard/sam/AbstractAlignmentMerger.java]: in this code that you removed calculateNm, calculateMd were reversed!

Comment (nh13) [inline src/main/java/picard/sam/AbstractAlignmentMerger.java]: I see it.  I really like scala where we can have named params.  Sigh.

**Comment by yfarjoun** [comment]
if I understand correctly, this reverts the fixing of tags when the cigar changes, but includes MD in the fixTags CLP. Should we purge MD NM and UQ  when we change the cigar? At-least we will not be leaving around incorrect information....

File: src/main/java/picard/sam/AbstractAlignmentMerger.java (hunk 1 above)
Comment (yfarjoun) [inline src/main/java/picard/sam/AbstractAlignmentMerger.java]: I agree. especially important for booleans. but all params could benefit. not sure why java don't have it too....

**Comment by nh13** [comment]
@yfarjoun: If downstream users are impacted (i.e. they want MD tags but no longer have them), then they have been relying on potentially the wrong values.  I vote for removing NM/MD/UQ when we clip overlapping reads.

Also, should we do the same for clipping adapters (see `updateCigarForTrimmedOrClippedBases` and `this.clipAdapters`)?

File: src/main/java/picard/sam/SetNmMDAndUqTags.java [hunk 2]
```diff
@@ -48,19 +47,19 @@
...
-        usage = SetNmAndUqTags.USAGE_SUMMARY + SetNmAndUqTags.USAGE_DETAILS,
-        usageShort = SetNmAndUqTags.USAGE_SUMMARY,
+        usage = SetNmMDAndUqTags.USAGE_SUMMARY + SetNmMDAndUqTags.USAGE_DETAILS,
+        usageShort = SetNmMDAndUqTags.USAGE_SUMMARY,
         programGroup = SamOrBam.class
 )
-public class SetNmAndUqTags extends CommandLineProgram {
-    static final String USAGE_SUMMARY = "Fixes the UQ and NM tags in a SAM file.  ";
-    static final String USAGE_DETAILS = "This tool takes in a SAM or BAM file (sorted by coordinate) and calculates the NM and UQ tags by comparing with the reference."+
+public class SetNmMDAndUqTags extends CommandLineProgram {
```
Comment (tfenne) [inline src/main/java/picard/sam/SetNmMDAndUqTags.java]: It's going to drive me crazy that it's `Nm`, `Uq` and `MD` in the name.  Can you change to `Md` please?

**Comment by tfenne** [comment]
@nh13 @yfarjoun Agreed that MergeBamAlignment should probably remove tags rather than output invalid values.  The other option is to add back @nh13's change _but_ have it use either the indexed fasta file (potentially loading the same region many times) or load the entire reference into memory (requiring more memory).

**Comment by yfarjoun** [comment]
that would require a new type of ReferenceFile reader, no?

**Comment by nh13** [comment]
@yfarjoun I think I will remove the tags, and in the case folks output queryname, the old behavior stands.  I will add some documentation somewhere about this too.

**Comment by yfarjoun** [comment]
but even if outputting in queryname the tags will be incorrect...so why not
remove them from the reads that are clipped?

**Comment by nh13** [comment]
@yfarjoun see my latest changes.  I remove the NM/MD/UQ tags for reads that are clipped due to adapters or overlapping bases.  I also added some doc to `MergeBamAlignment`.

File: src/main/java/picard/sam/AbstractAlignmentMerger.java [hunk 3]
```diff
@@ -477,7 +476,7 @@ public void mergeAlignment(final File referenceFasta) {
 
             for (final SAMRecord rec : sink.sorter) {
                 if (!rec.getReadUnmappedFlag() && refSeq != null) {
-                    fixNMandUQ(rec, refSeq, bisulfiteSequence);
+                    fixNmMdAndUQ(rec, refSeq, bisulfiteSequence);
```
Comment (yfarjoun) [inline src/main/java/picard/sam/AbstractAlignmentMerger.java]: UQ -> Uq

File: src/main/java/picard/sam/AbstractAlignmentMerger.java [hunk 4]
```diff
@@ -499,10 +498,13 @@ public void mergeAlignment(final File referenceFasta) {
      *
      * No return value, modifies the provided record.
      */
-    public static void fixNMandUQ(final SAMRecord record, final ReferenceSequenceFileWalker refSeqWalker, final boolean isBisulfiteSequence) {
+    public static void fixNmMdAndUQ(final SAMRecord record, final ReferenceSequenceFileWalker refSeqWalker, final boolean isBisulfiteSequence) {
```
Comment (yfarjoun) [inline src/main/java/picard/sam/AbstractAlignmentMerger.java]: Uq

File: src/main/java/picard/sam/AbstractAlignmentMerger.java [hunk 5]
```diff
@@ -619,13 +607,13 @@ protected void clipForOverlappingReads(final SAMRecord read1, final SAMRecord re
                     if (posDiff > 0) {
                         CigarUtil.softClip3PrimeEndOfRead(pos, Math.min(pos.getReadLength(),
                                 pos.getReadLength() - posDiff + 1));
-                        calculateMdAndNmTags(pos);
+                        removeNmMdAndUqTags(pos); // these tags are no invalid!
```
Comment (yfarjoun) [inline src/main/java/picard/sam/AbstractAlignmentMerger.java]: no->not

File: src/main/java/picard/sam/AbstractAlignmentMerger.java [hunk 6]
```diff
@@ -619,13 +607,13 @@ protected void clipForOverlappingReads(final SAMRecord read1, final SAMRecord re
...
                     }
 
                     if (negDiff > 0) {
                         CigarUtil.softClip3PrimeEndOfRead(neg, Math.min(neg.getReadLength(),
                                 neg.getReadLength() - negDiff + 1));
-                        calculateMdAndNmTags(neg);
+                        removeNmMdAndUqTags(neg); // these tags are no invalid!
```
Comment (yfarjoun) [inline src/main/java/picard/sam/AbstractAlignmentMerger.java]: not

File: src/main/java/picard/sam/AbstractAlignmentMerger.java [hunk 7]
```diff
@@ -752,6 +740,7 @@ protected void updateCigarForTrimmedOrClippedBases(final SAMRecord rec, final SA
         // If the adapter sequence is marked and clipAdapter is true, clip it
         if (this.clipAdapters && rec.getAttribute(ReservedTagConstants.XT) != null) {
             CigarUtil.softClip3PrimeEndOfRead(rec, rec.getIntegerAttribute(ReservedTagConstants.XT));
+            removeNmMdAndUqTags(rec); // these tags are no invalid!
```
Comment (yfarjoun) [inline src/main/java/picard/sam/AbstractAlignmentMerger.java]: not

File: src/main/java/picard/sam/AbstractAlignmentMerger.java [hunk 8]
```diff
@@ -813,4 +802,15 @@ public void setIncludeSecondaryAlignments(final boolean includeSecondaryAlignmen
     public void close() {
         CloserUtil.close(this.refSeq);
     }
+
+
+    /** Removes the NM, MD, and UQ tags.  This is useful if we modify the read and are not able to recompute these tags,
+     * for example when no reference is available.
+     * @param rec the record to modify.
+     */
+    private void removeNmMdAndUqTags(final SAMRecord rec) {
```
Comment (yfarjoun) [inline src/main/java/picard/sam/AbstractAlignmentMerger.java]: can be static

File: src/main/java/picard/sam/AbstractAlignmentMerger.java (hunk 5 above)
Comment (yfarjoun) [inline src/main/java/picard/sam/AbstractAlignmentMerger.java]: In the case where the read is already clipped, the tags might be correct. Should we compare new-vs-old CIGAR and only remove tags if there's a difference?

File: src/main/java/picard/sam/SetNmMDAndUqTags.java [hunk 9]
```diff
@@ -48,19 +47,19 @@
...
-        usage = SetNmAndUqTags.USAGE_SUMMARY + SetNmAndUqTags.USAGE_DETAILS,
-        usageShort = SetNmAndUqTags.USAGE_SUMMARY,
+        usage = SetNmMdAndUqTags.USAGE_SUMMARY + SetNmMdAndUqTags.USAGE_DETAILS,
+        usageShort = SetNmMdAndUqTags.USAGE_SUMMARY,
         programGroup = SamOrBam.class
 )
-public class SetNmAndUqTags extends CommandLineProgram {
-    static final String USAGE_SUMMARY = "Fixes the UQ and NM tags in a SAM file.  ";
-    static final String USAGE_DETAILS = "This tool takes in a SAM or BAM file (sorted by coordinate) and calculates the NM and UQ tags by comparing with the reference."+
+public class SetNmMdAndUqTags extends CommandLineProgram {
```
Comment (yfarjoun) [inline src/main/java/picard/sam/SetNmMDAndUqTags.java]: This is a breaking change...can we have a deprecated version of the old class-name? for a few releases?

**Comment by yfarjoun** [comment]
can I haz  A small test that shows that the tags are actually removed?

**Comment by yfarjoun** [comment]
back to you @nh13

**Comment by yfarjoun** [comment]
why assigning me here? this has some work to be done:

- tests are not passing
- need to add a small test
- add a deprecated wrapper with the old name?

Thanks! :-)

**Comment by nh13** [comment]
@yfarjoun you need to un-assign yourself even if you assign someone else.  Multiple folks can be assigned an issue now.

File: src/main/java/picard/sam/AbstractAlignmentMerger.java (hunk 3 above)
Comment (nh13) [inline src/main/java/picard/sam/AbstractAlignmentMerger.java]: Fixed.

File: src/main/java/picard/sam/AbstractAlignmentMerger.java (hunk 4 above)
Comment (nh13) [inline src/main/java/picard/sam/AbstractAlignmentMerger.java]: Fixed.

File: src/main/java/picard/sam/AbstractAlignmentMerger.java (hunk 5 above)
Comment (nh13) [inline src/main/java/picard/sam/AbstractAlignmentMerger.java]: I think this a lot of effort for little reward, since likely one would need to run `SetNmMdAndUqTags` anyhow.

File: src/main/java/picard/sam/AbstractAlignmentMerger.java (hunk 6 above)
Comment (nh13) [inline src/main/java/picard/sam/AbstractAlignmentMerger.java]: `now` :)

File: src/main/java/picard/sam/AbstractAlignmentMerger.java (hunk 7 above)
Comment (nh13) [inline src/main/java/picard/sam/AbstractAlignmentMerger.java]: Fixed.

File: src/main/java/picard/sam/AbstractAlignmentMerger.java (hunk 8 above)
Comment (nh13) [inline src/main/java/picard/sam/AbstractAlignmentMerger.java]: It's private, but sure.

File: src/main/java/picard/sam/SetNmMDAndUqTags.java (hunk 9 above)
Comment (nh13) [inline src/main/java/picard/sam/SetNmMDAndUqTags.java]: Done.

**Comment by nh13** [comment]
@yfarjoun added the test and rebased onto master.  After a final 👍 this is ready to go.

File: src/main/java/picard/sam/SetNmMdAndUqTags.java [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+    protected String[] customCommandLineValidation() {
+        if (REFERENCE_SEQUENCE == null) {
+            return new String[]{"Must have a non-null REFERENCE_SEQUENCE"};
+        }
+        return super.customCommandLineValidation();
+    }
+
+    private final Log log = Log.getInstance(SetNmMdAndUqTags.class);
+
+    public static void main(final String[] argv) {
```
Comment (yfarjoun) [inline src/main/java/picard/sam/SetNmMdAndUqTags.java]: is this main still needed with the new system?

File: src/main/java/picard/sam/SetNmMdAndUqTags.java [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+            throw new SAMException("Input must be coordinate-sorted for this program to run. Found: " + reader.getFileHeader().getSortOrder());
+        }
+
+        final SAMFileWriter writer = new SAMFileWriterFactory().makeSAMOrBAMWriter(reader.getFileHeader(), true, OUTPUT);
+        writer.setProgressLogger(
+                new ProgressLogger(log, (int) 1e7, "Wrote", "records"));
+
+        final ReferenceSequenceFileWalker refSeq = new ReferenceSequenceFileWalker(REFERENCE_SEQUENCE);
+
+       StreamSupport.stream(reader.spliterator(),false)
```
Comment (yfarjoun) [inline src/main/java/picard/sam/SetNmMdAndUqTags.java]: extra space missing here?

Comment (yfarjoun) [inline src/main/java/picard/sam/SetNmMdAndUqTags.java]: aligning the line, and after comma.

**Comment by yfarjoun** [comment]
sorry for the delay. thanks!! 👍

File: src/main/java/picard/sam/SetNmMdAndUqTags.java (hunk 10 above)
Comment (nh13) [inline src/main/java/picard/sam/SetNmMdAndUqTags.java]: Not sure, but I am going to leave it here as other classes have it; cleaning this up should be in all classes could be in another PR.

File: src/main/java/picard/sam/SetNmMdAndUqTags.java (hunk 11 above)
Comment (nh13) [inline src/main/java/picard/sam/SetNmMdAndUqTags.java]: Fixed
